devicetree: Add DT_NODELABEL_C_TOKEN_* macros

rruuaanng opened the pull request:
Add the DT_NODELABEL_C_TOKEN and
DT_NODELABEL_C_TOKEN_BY_IDX macros to retrieve
the C symbolic name of a node.

Closes: #67046

decsny commented on scripts/dts/gen_defines.py:
```diff
@@ -815,11 +815,13 @@ def phandle_macros(prop: edtlib.Property, macro: str) -> dict:
         for i, node in enumerate(prop.val):
             ret[f"{macro}_IDX_{i}"] = f"DT_{node.z_path_id}"
             ret[f"{macro}_IDX_{i}_PH"] = f"DT_{node.z_path_id}"
+            ret[f"{macro}_IDX_{i}_PH_TOKEN"] = f"{node.labels[0]}"
```
isn't this going to get the first nodelabel found on the node and not necessarily the token in the phandle specifier?

rruuaanng replied:
Since a node can have multiple labels, e.g.:
```
foo1: foo2: bar {
  ...
} 
```
Then, through statement `node.labels[0]`, it will obtain `foo1`. Although only a few users would write it this way, regardless, only the leftmost label will be generated because the node expansion syntax is as follows:
```
label [:...] ... : nodename {
  ...
}
```

decsny replied:
yeah, this is exactly what I mean, is that, if you have a node like:

```
foo1: foo2: foo {
  ...
};

bar {
   some-phandle = <&foo2>; 
};
```

then this macro `DT_PHANDLE_TOKEN(bar, some_phandle)` is going to return `foo1` when it seems like from user perspective they want to get `foo2`

rruuaanng replied:
Oh, that's great. It does exist in reality, but solving it is a bit tricky. The challenge lies in how to distinguish between the generated macros.

rruuaanng replied:
@decsny If an additional macro can be tolerated, then the issue could be resolved. It might seem a bit odd since this API processes multiple nodes, but honestly, I think it might be better to keep things as they are, because in practice, using multiple tags is generally discouraged. WDYT?

```
#define DT_PHANDLE_TOKEN_BY_LABEL_IDX(node_id, prop, idx, label_idx) \
	DT_CAT6(node_id, _P_, prop, _IDX_, idx, _PH_TOKEN_, label_idx)

#define DT_N_S_test_S_phandle_holder_0_P_phs_IDX_1_PH_TOKEN_0 test_spi1
#define DT_N_S_test_S_phandle_holder_0_P_phs_IDX_1_PH_TOKEN_1 test_spi2
```

decsny reviewed (changes requested):
need to sort out the expectation of this macro

rruuaanng commented:
When defining multiple labels in a phandle node, use `DT_PHANDLE_TOKEN_BY_
LABEL_IDX` to specify the label index, while `DT_PHANDLE_TOKEN_BY_IDX` is used for scenarios involving a single label.

mbolivar commented:
Do we still need this? I note that the original issue, https://github.com/zephyrproject-rtos/zephyr/issues/67046, was opened for the ubxlib project, which is not being actively developed anymore: https://github.com/u-blox/ubxlib/blob/master/README.md

Do we have any other use cases for this?

rruuaanng commented:
Yeah, it seems like the project has been stopped. Personally, I believe that when this requirement was officially raised, there was definitely an audience for it, even though no one has paid attention to the original issue for a long time. I think this work is forward-thinking, at least we’ve provided this feature, so users don’t have to attempt opening an issue upstream.

mbolivar reviewed (changes requested):
@rruuaanng fair enough! Here is a quick review, then. Thanks for taking the time to work on this.

kartben commented on scripts/dts/gen_defines.py:
```diff
@@ -815,11 +815,19 @@ def phandle_macros(prop: edtlib.Property, macro: str) -> dict:
         ret[f"{macro}"] = f"DT_{prop.val.z_path_id}"
         ret[f"{macro}_IDX_0"] = f"DT_{prop.val.z_path_id}"
         ret[f"{macro}_IDX_0_PH"] = f"DT_{prop.val.z_path_id}"
+        ret.update({
```
Pretty sure there are some gaps there over the past 12-18 months or so btw

E.g.
https://github.com/zephyrproject-rtos/zephyr/commit/4553a21fe52001592c2b96ff4c6c8e673ddd950f

rruuaanng replied:
I just checked file `macros.bnf`, and it seems there are many missing entries. I will follow up on it when I have more time.

rruuaanng commented on include/zephyr/devicetree.h:
```diff
@@ -1798,6 +1798,73 @@
 #define DT_PHANDLE(node_id, prop) DT_PHANDLE_BY_IDX(node_id, prop, 0)
 
+/**
+ * @brief Get the symbolic name of a phandle by position and label index
```
Naming is really difficult. What do you think of `DT_PHANDLE_C_TOKEN`?

rruuaanng commented on include/zephyr/devicetree.h:
```diff
@@ -1798,6 +1798,73 @@
+ *             with type `phandle`
+ * @return symbolic name of the phandle
+ */
+#define DT_PHANDLE_TOKEN(node_id, prop) \
```
Maybe not. I believe keeping `node_id` is the simplest and most straightforward approach, at least based on past real-world use cases.

rruuaanng commented:
I renamed it to `DT_NODELABEL_C_TOKEN` so that we can reuse other macros related to node identifiers and obtain its C token. e.g., the test I provided.

rruuaanng commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -367,6 +367,37 @@ ZTEST(devicetree_api, test_has_nodelabel)
+	zassert_equal(test_spi, "test_spi");
+
+#undef TEST_PHS_IDX
+#undef TEST_GPIO_STR
```
```suggestion
```

kartben reviewed (dismissed):
release note update needs to be made against 4.3 file

tomi-font commented:
@mbolivar please review or unassign yourself

kartben commented on scripts/dts/gen_defines.py:
```diff
@@ -245,6 +245,10 @@ def write_idents_and_existence(node: edtlib.Node) -> None:
 
+    # Node label tokens
+    for i, label in enumerate(node.labels[::-1]):
+        out_dt_define(f"N_NODELABEL_DT_{node.z_path_id}_IDX_{i}_C_TOKEN", label)
```
need to update `doc/build/dts/macros.bnf` please

rruuaanng commented on include/zephyr/devicetree.h:
```diff
@@ -195,6 +195,56 @@
+ *
+ * Example usage:
+ *
+ * @code{.c}
+ *     #define N1 DT_NODELABEL(n1)
+ *
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 0), 0) // n2_1
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 1), 0) // n3_1
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 0), 1) // n2
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 1), 1) // n3
+ * @endcode
```
It relies on the correctness of `DT_PHANDLE_BY_IDX`. Of course, `DT_PHANDLE_BY_IDX` has been extensively tested and used, and there haven’t been any reported issues with retrieving phandles in scenarios involving multiple overlays. Therefore, I believe it is correct.

It seems I should clarify the index order, in the example, the index appears to increase from right to left.

rruuaanng replied:
Ah, I added a `details` section to make things clearer for developers :)

jhedberg commented:
@rruuaanng this needs a rebase due to a merge conflict

rruuaanng commented on scripts/dts/gen_defines.py:
```diff
@@ -3,7 +3,7 @@
 # Copyright (c) 2019 Linaro Limited
 # Copyright (c) 2024 SILA Embedded Solutions GmbH
-# SPDX-License-Identifier: BSD-3-Clause
+# SPDX-License-Identifier: Apache-2.0
```
I have removed that change.

mbolivar commented:
@decsny ping on this one

pdgendt commented on tests/lib/devicetree/api/app.overlay:
```diff
@@ -231,7 +231,7 @@
 		};
 
-		test_i2c: i2c@11112222 {
+		test_i2c: test_i2c1: i2c@11112222 {
```
I see @decsny made the same [comment](https://github.com/zephyrproject-rtos/zephyr/pull/86973#pullrequestreview-3245740273) before.

If it's not an issue, ignore this comment.

rruuaanng replied:
I have modified that example :)

pdgendt commented on include/zephyr/devicetree.h:
```diff
@@ -196,6 +196,57 @@
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 0), 0) // n2
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 0), 1) // n2_1
+ *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 1), 0) // n2
+ * 	   DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 1), 1) // n2_1
```
Random tab character
```suggestion
 *     DT_NODELABEL_C_TOKEN_BY_IDX(DT_PHANDLE_BY_IDX(N1, foo, 1), 1) // n2_1
```

jhedberg commented:
@rruuaanng needs a rebase, and you need to switch to `release-notes-4.5.rst` since this missed 4.4.

rruuaanng committed:
devicetree: Add DT_NODELABEL_C_TOKEN_* macros

Add the `DT_NODELABEL_C_TOKEN` and
`DT_NODELABEL_C_TOKEN_BY_IDX` macros to retrieve
the C symbolic name of a node.

Signed-off-by: James Roy <rruuaanng@outlook.com>

mbolivar commented:
> @decsny ping on this one

No response since December, dismissing this -1 as DT maintainer.
